subprojects/build-operations/src/main/java/org/gradle/internal/operations/DefaultBuildOperationRunner.java

```diff
@@ -0,0 +1,354 @@
+                setCurrentBuildOperation(parentOperation);
+                operationState.setRunning(false);
+                LOGGER.debug("Build operation '{}' completed", descriptor.getDisplayName());
+            }
+        });
+    }
+
+    @OverridingMethodsMustInvokeSuper
+    protected <O> O execute(BuildOperationDescriptor descriptor, BuildOperationState operationState, BuildOperationExecution<O> execution, BuildOperationExecutionListener listener) {
+        return execution.execute(
+            descriptor,
+            operationState,
+            new DefaultBuildOperationContext(),
+            listener
+        );
+    }
+
+    private void assertParentRunning(String message, BuildOperationDescriptor child, @Nullable BuildOperationState parent) {
+        if (parent != null && !parent.isRunning()) {
+            String parentName = parent.getDescription().getDisplayName();
+            throw new IllegalStateException(String.format(message, child.getDisplayName(), parentName));
+        }
+    }
+
+    @Override
+    public BuildOperationRef getCurrentOperation() {
+        BuildOperationRef current = getCurrentBuildOperation();
+        if (current == null) {
+            throw new IllegalStateException("No operation is currently running.");
+        }
+        return current;
+    }
+
+    @Nullable
+    protected BuildOperationState getCurrentBuildOperation() {
+        return (BuildOperationState) currentBuildOperationRef.get();
+    }
+
+    protected void setCurrentBuildOperation(@Nullable BuildOperationState parentState) {
+        currentBuildOperationRef.set(parentState);
+    }
+
+    @Nullable
+    private static BuildOperationState determineParent(BuildOperationDescriptor.Builder descriptorBuilder, @Nullable BuildOperationState defaultParent) {
+        BuildOperationState parent = (BuildOperationState) descriptorBuilder.getParentState();
+        if (parent == null) {
+            parent = defaultParent;
+        }
+        return parent;
+    }
+
+    @OverridingMethodsMustInvokeSuper
+    protected BuildOperationDescriptor createDescriptor(BuildOperationDescriptor.Builder descriptorBuilder, @Nullable BuildOperationState parent) {
+        OperationIdentifier id = new OperationIdentifier(buildOperationIdFactory.nextId());
+        return descriptorBuilder.build(id, parent == null
+            ? null
+            : parent.getDescription().getId());
+    }
+
+    public interface TimeSupplier {
+        long getCurrentTime();
+    }
+
+    private static RuntimeException throwAsUncheckedException(Throwable t) {
+        if (t instanceof InterruptedException) {
+            Thread.currentThread().interrupt();
+        }
+        if (t instanceof RuntimeException) {
+            throw (RuntimeException) t;
+        }
+        if (t instanceof Error) {
+            throw (Error) t;
+        }
+        if (t instanceof IOException) {
+            throw new RuntimeException(t.getMessage(), t);
```

Comment:
Why is here the extra branch for IOException? Deleting this `if` should yield the same result, right?

Reply:
Ah, right. We used to convert this to `UncheckedIOException`, but that doesn't work with Java 6... :/

Reply:
Fixed.